### PR TITLE
NLPGeneratorAnalyzer extended for text attributes

### PR DESCRIPTION
**Pull request by amrasekh**
## Change description
Extended NLPGeneratorAnalyzer for flexible text attributes assessment
Debugged data paths
Upgraded the demo

## Type of change
- [x] Bug fix (fixes an issue)
- [x] New feature (adds functionality)

## Related issues

> Fix [#1]() 

### Development

- [ ] Lint rules pass locally
- [ ] Application changes have been tested thoroughly
- [ ] Automated tests covering modified code pass

### Security

- [ ] Security impact of change has been considered
- [ ] Code follows company security practices and guidelines

### Code review 

- [ ] Pull request has a descriptive title and context useful to a reviewer. Screenshots or screencasts are attached as necessary
- [ ] "Ready for review" label attached and reviewers assigned
- [ ] Changes have been reviewed by at least one other contributor
- [ ] Pull request linked to task tracker where applicable

### CODE REVIEW

**Comment by IanAtCredo** [comment]
The `tensorflow hub` requirement for the demo is very heavy, just for the encoder. It is necessary for the pretrained model I take it, but it'd be great if we could use an encoder that's part of the transformers package itself.

On M1 Mac it took a while to figure out how to install as well. Was failing for a while.

**Comment by IanAtCredo** [comment]
`assessment_fun` and `assessment_attribute` could probably be combined into one argument that's a dictionary.

```
{assessment_attribute1: fun1,
assessment_attribute2: fun2,
...}
```

Refactor the module so that it runs the multiple assessments.

**Comment by IanAtCredo** [comment]
There will be humongous time improvements if the tokenized prompts can be passed to gpt2 as a batch. Doing this the `generator_fun` should take in a array of prompts rather than a single prompt.

I think we should move there. I tried to do that, but sequences of varying lengths don't work. Some padding needs to be done, but not how to do that with gpt2.

**Review by IanAtCredo — CHANGES_REQUESTED** [review]
Ok, went through and made some minor changes. Issues I have:

* installation is difficult on m1 because of tensorflow. Is it possible to remove that dependency? Not terrible if not, but not perfect.
* Generator function should take in "prompts" rather than a single prompt. Massive speed ups are possible and right now it is very slow.
* Would be great for this to take in a set of assessment functions that can be run. That way the generation can be done _once_ (which will be quicker if the above is implemented) for multiple assessments. Right now it just takes way too long.

**Comment by amrasekh** [comment]
> Ok, went through and made some minor changes. Issues I have:
> 
> * installation is difficult on m1 because of tensorflow. Is it possible to remove that dependency? Not terrible if not, but not perfect.
> * Generator function should take in "prompts" rather than a single prompt. Massive speed ups are possible and right now it is very slow.
> * Would be great for this to take in a set of assessment functions that can be run. That way the generation can be done _once_ (which will be quicker if the above is implemented) for multiple assessments. Right now it just takes way too long.

Item 3 is now implemented. Ticket DS-278 is created to cover 1 and 2 in a next PR and added to backlog.

**Review by IanAtCredo — APPROVED** [review]
